crates/wasmparser/src/readers/core/memories.rs

```diff
@@ -32,23 +32,40 @@ impl<'a> FromReader<'a> for MemoryType {
         let has_max = flags & 0b0001 != 0;
         let has_page_size = flags & 0b1000 != 0;
 
+        if memory64 && !reader.memory64() {
+            bail!(pos, "memory64 must be enabled for 64-bit memories");
+        }
+        if shared && !reader.threads() {
+            bail!(pos, "threads must be enabled for shared memories");
+        }
+        if has_page_size && !reader.custom_page_sizes() {
+            bail!(
+                pos,
+                "the custom page sizes proposal must be enabled to customize a memory's page size"
+            );
+        }
```

Comment:
Could these feature checks be moved back to the validator? Locally it looks like `tests/cli/missing-features/custom-page-sizes.wast` fails but that can be switched to `assert_invalid` from `assert_malformed`.  

Or otherwise, more generally, I'd prefer to keep feature flag checking entirely in the validator where possible. If a local test is the one forcing a feature flag check then it's definitely ok to change local tests to `assert_invalid` instead of `assert_malformed`.

Reply:
Got it. I was able to move two out of the three -- the `shared` check is currently required by the spec `binary.wast` test but can get moved to the validator once the wasm-3.0 branch is merged to main in the spec repo.